### PR TITLE
Test versions mismatch

### PR DESCRIPTION
**Pull request by a-masterov**
## Problem
We faced the problem of incompatibility of the different components of different versions.
This should be detected automatically to prevent production bugs.
## Summary of changes
The test for this situation was implemented

## Checklist before requesting a review

- [x] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

File: test_runner/regress/test_compatibility.py
Comment (bayandin) [inline test_runner/regress/test_compatibility.py:646]: I'd try not to mix `pathlib`'s API with `os`' counterparts, let's use [`pathlib.Path.mkdir`](https://docs.python.org/3/library/pathlib.html#pathlib.Path.mkdir) instead, it should be something like:
```python
target_dir.mkdir(parents=True)
```

File: test_runner/regress/test_compatibility.py
Comment (a-masterov) [reply test_runner/regress/test_compatibility.py]: Fixed, thanks

File: test_runner/regress/test_storage_controller.py
Comment (bayandin) [inline test_runner/regress/test_storage_controller.py:87]: Let's try to make it possible to apply parameterization without changing the test itself (otherwise it's going to be a massive copy-paste). 
Afiar, you can access `request.param` in `neon_env_builder` fixture to get current parameter

File: test_runner/regress/test_storage_controller.py
Comment (bayandin) [inline test_runner/regress/test_storage_controller.py:61]: It seems the combination with hexes looks pretty unreadable 🤔 
I'd create a separate function that would return this list and then add a comment with an explanation of the function.

Or even better, create a function that we'll call inside `@pytest.mark.parametrize`, something like:

```python
@pytest.mark.parametrize(some_function_name())
def test_smth(...):
    ...
```

File: test_runner/fixtures/neon_fixtures.py
Comment (bayandin) [inline test_runner/fixtures/neon_fixtures.py:497]: Probably we can move it to "constants"

File: test_runner/fixtures/neon_fixtures.py
Comment (bayandin) [inline test_runner/fixtures/neon_fixtures.py:508]: And this to a separate function

File: test_runner/fixtures/neon_fixtures.py
Comment (bayandin) [inline test_runner/fixtures/neon_fixtures.py:466]:
```suggestion
        self.mixdir = self.test_output_dir / "mixdir_neon"
```

**Review by bayandin — COMMENTED** [review]
It already looks awesome!
I've added several comments regarding making it more user-friendly and to hide internal implementation details.

Good work!

File: test_runner/regress/test_storage_controller.py
Comment (a-masterov) [reply test_runner/regress/test_storage_controller.py]: I believe this approach is implemented now

File: test_runner/fixtures/neon_fixtures.py
Comment (a-masterov) [reply test_runner/fixtures/neon_fixtures.py]: done

File: test_runner/fixtures/neon_fixtures.py
Comment (a-masterov) [reply test_runner/fixtures/neon_fixtures.py]: done

File: test_runner/fixtures/neon_fixtures.py
Comment (bayandin) [inline test_runner/fixtures/neon_fixtures.py:130]: Could you add more verbose comment here, please

File: test_runner/fixtures/neon_fixtures.py
Comment (bayandin) [inline test_runner/fixtures/neon_fixtures.py:177]: I don't think this should be a fixture, it can be a regular helper-function

File: test_runner/fixtures/neon_fixtures.py
Comment (bayandin) [inline test_runner/fixtures/neon_fixtures.py:188]: I don't think this should fall back to a hardcoded `neon_previous`. I'd say it just should fail, and ask to set `COMPATIBILITY_NEON_BIN` path

File: test_runner/fixtures/neon_fixtures.py
Comment (bayandin) [inline test_runner/fixtures/neon_fixtures.py:184]: I don't think this check for `pageserver` is required here

File: test_runner/fixtures/neon_fixtures.py
Comment (bayandin) [inline test_runner/fixtures/neon_fixtures.py:431]: It doesn't seem that `request` belong to `NeonEnvBuilder` class.
`neon_env_builder` fixture should parse `combination` parameter and create `NeonEnvBuilder` with `version_combination` argument.

File: test_runner/fixtures/neon_fixtures.py
Comment (bayandin) [inline test_runner/fixtures/neon_fixtures.py:737]: Let's use [`Path.symlink_to`](https://docs.python.org/3/library/pathlib.html#pathlib.Path.symlink_to) here

File: test_runner/fixtures/neon_fixtures.py
Comment (bayandin) [inline test_runner/fixtures/neon_fixtures.py:212]: The same comment as for `compatibility_neon_binpath`.

File: test_runner/fixtures/neon_fixtures.py
Comment (a-masterov) [reply test_runner/fixtures/neon_fixtures.py]: this code was refactored

File: test_runner/fixtures/neon_fixtures.py
Comment (a-masterov) [reply test_runner/fixtures/neon_fixtures.py]: This code was removed

File: test_runner/regress/test_storage_controller.py
Comment (a-masterov) [reply test_runner/regress/test_storage_controller.py]: done

File: test_runner/fixtures/neon_fixtures.py
Comment (a-masterov) [reply test_runner/fixtures/neon_fixtures.py]: fixed, thanks

File: test_runner/fixtures/neon_fixtures.py
Comment (a-masterov) [reply test_runner/fixtures/neon_fixtures.py]: done

File: test_runner/fixtures/neon_fixtures.py
Comment (a-masterov) [reply test_runner/fixtures/neon_fixtures.py]: removed

File: test_runner/fixtures/utils.py
Comment (bayandin) [inline test_runner/fixtures/utils.py:631]: If we're moving away from bitmasks, I think it makes sense to use use "old" and "new" — this will improve readability a lot

File: test_runner/regress/test_compatibility.py
Comment (bayandin) [inline test_runner/regress/test_compatibility.py:585]: Ideally, It would be nice to build-in all `parameterize`parameters into one function:
I.e. instead of 
```python
@pytest.mark.parametrize("combination", all_pairs_component_versions(), ids=fixtures.utils.comb_ids)
``` 

make it 
```python
@pytest.mark.parametrize(all_pairs_component_versions())
# or, probably it'll be 
@pytest.mark.parametrize(**all_pairs_component_versions())
```

File: test_runner/fixtures/utils.py
Comment (bayandin) [inline test_runner/fixtures/utils.py:637]: We need to add a constraint that there's no "all new" and no "all old" combinations

File: test_runner/regress/test_compatibility.py
Comment (a-masterov) [reply test_runner/regress/test_compatibility.py]: done

File: test_runner/fixtures/utils.py
Comment (a-masterov) [reply test_runner/fixtures/utils.py:641]: But we need the "all-new" combination to test if it works on new versions

File: test_runner/fixtures/neon_fixtures.py
Comment (a-masterov) [reply test_runner/fixtures/neon_fixtures.py]: done

File: test_runner/fixtures/utils.py
Comment (a-masterov) [reply test_runner/fixtures/utils.py:631]: done

File: test_runner/fixtures/neon_fixtures.py
Comment (bayandin) [inline test_runner/fixtures/neon_fixtures.py:450]: Probably should be done in the `neon_binpath` fixture

File: test_runner/fixtures/neon_fixtures.py
Comment (bayandin) [inline test_runner/fixtures/neon_fixtures.py:711]: Hmm, in which case we could have this file existing?

File: test_runner/fixtures/neon_fixtures.py
Comment (bayandin) [inline test_runner/fixtures/neon_fixtures.py:1503]: We need to move this expression to a separate variable as well.

File: test_runner/fixtures/utils.py
Comment (bayandin) [reply test_runner/fixtures/utils.py:631]: The comment still says False / True; should it also be updated?

File: test_runner/fixtures/neon_fixtures.py
Comment (a-masterov) [reply test_runner/fixtures/neon_fixtures.py]: it's not needed at all

File: test_runner/fixtures/neon_fixtures.py
Comment (a-masterov) [reply test_runner/fixtures/neon_fixtures.py]: We couldn't, removed

File: test_runner/fixtures/neon_fixtures.py
Comment (a-masterov) [reply test_runner/fixtures/neon_fixtures.py]: done, thanks

File: test_runner/fixtures/utils.py
Comment (bayandin) [reply test_runner/fixtures/utils.py]: Yeah, you're right, we need to ensure that combinations **must** include "all new", and **must not** include "all old"

Comment (a-masterov) [reply test_runner/fixtures/utils.py]: done

File: test_runner/fixtures/utils.py
Comment (bayandin) [inline test_runner/fixtures/utils.py:39]: The order of imports became weird after the latest merge.

- `from collections import OrderedDict` should go to the top block of system imports
- `from allpairspy import AllPairs` should to the second block with imports from dependencies
- also, if you could move the line below (`from fixtures.common_types import TimelineId`) to the third block of local imports  that would be nice

File: test_runner/fixtures/utils.py
Comment (bayandin) [inline test_runner/fixtures/utils.py:647]: `AllPairs` accepts `filter_func` to filter out unwanted combinations (see the [README](https://github.com/thombashi/allpairspy?tab=readme-ov-file)).
I would filter out all-new and all-old combinations and then add all-new to the returning result.

Comment (a-masterov) [reply test_runner/fixtures/utils.py:646]: If I use the filter function it generates one more combination.
So without the filter, we had 6 combinations, but if we filter out the all-new combination we still get 6, so it will be 7 combinations then if we add all-new one by hand

File: test_runner/fixtures/utils.py
Comment (a-masterov) [reply test_runner/fixtures/utils.py]: Fixed, thanks

File: test_runner/fixtures/utils.py
Comment (bayandin) [inline test_runner/fixtures/utils.py:66]: Let's keep this as a function and generate `ids` automatically to make it easier to add more combinations.
Also, let's add a comment with an explanation of what's going on here (the function generates parameters for `@pytest.mark.parametrize`, why we choose only this set of parameters, etc)

Comment (a-masterov) [reply test_runner/fixtures/utils.py]: done

File: test_runner/fixtures/neon_fixtures.py
Comment (bayandin) [inline test_runner/fixtures/neon_fixtures.py:496]: Aren't we moving `.absolute()` to the fixture?

File: test_runner/fixtures/utils.py
Comment (bayandin) [inline test_runner/fixtures/utils.py:91]: The previous version had better readability, let's bring it back.
We can let auto-formatter to accept if this is the problem like this: https://github.com/neondatabase/neon/blob/c2623ffef454378b2602f494e459b32028aa04a0/test_runner/performance/test_perf_olap.py#L107-L125

File: test_runner/fixtures/utils.py
Comment (bayandin) [inline test_runner/fixtures/utils.py:676]: Let's use f-strings here

File: test_runner/fixtures/utils.py
Comment (bayandin) [inline test_runner/fixtures/utils.py:48]: Let's use tuples instead of lists for `COMPONENT_BINARIES` and `VERSIONS_COMBINATIONS`

File: test_runner/fixtures/neon_fixtures.py
Comment (a-masterov) [reply test_runner/fixtures/neon_fixtures.py]: done

File: test_runner/fixtures/utils.py
Comment (a-masterov) [reply test_runner/fixtures/utils.py]: I got it, thanks

File: test_runner/fixtures/utils.py
Comment (a-masterov) [reply test_runner/fixtures/utils.py]: done

File: test_runner/fixtures/utils.py
Comment (a-masterov) [reply test_runner/fixtures/utils.py]: done, thanks

File: test_runner/fixtures/utils.py
Comment (bayandin) [inline test_runner/fixtures/utils.py:48]: You need to reenabled it back after the block

Comment (a-masterov) [reply test_runner/fixtures/utils.py:48]: done

File: test_runner/fixtures/utils.py
Comment (bayandin) [inline test_runner/fixtures/utils.py:641]: It seems ids is not set correctly, for example for `test_versions_mismatch[debug-pg17-combination` we have a bunch of the same `nnooo`:

Screenshot 2024-10-10 at 14 34 56

I guess `sorted` shouldn't be for values

**Review by bayandin — APPROVED** [review]
It should be ok to merge after the issues with ids are fixed.

Please ensure that it's convenient to run such tests locally (for example, all new combinations don't require old binaries).
Add a couple of lines about it to `test_runner/README.md`.
And please check with the Storage team that they're happy with the default set of combinations

Good work! 🎉

File: test_runner/fixtures/utils.py
Comment (a-masterov) [reply test_runner/fixtures/utils.py]: You are completely right.
Fixed, thanks!

**Review by bayandin — APPROVED** [review]

**Pull request merged**
